### ISSUE TITLE
User feedback: Tasker Integration

### ISSUE THREAD

**Issue by ignoramous** [celzero/brave-android-app#63]
Figure out what integration points are apt for Tasker

**Comment by Ch4t4r** [comment]
https://tasker.joaoapps.com/plugins-intro.html
You probably want to create an action plugin.

Essentially you "communicate" with tasker via Intents. There are two parts: An Activity where you configure this Intent (non-minimal example [here](https://git.frostnerd.com/PublicAndroidApps/smokescreen/-/blob/master/app/src/main/java/com/frostnerd/smokescreen/tasker/ConfigureActivity.kt)) and a BroadcastReceiver which receives an Intent previously configured (another ref [here](https://git.frostnerd.com/PublicAndroidApps/smokescreen/-/blob/master/app/src/main/java/com/frostnerd/smokescreen/tasker/FireReceiver.kt)).

**Comment by ignoramous** [comment]
See: #402

**Comment by Rhys-T** [comment]
If you end up making a Tasker plugin, it would probably make sense to have it check its caller for a Rethink-specific permission or something before accepting a command. Plugins can be called by any app on the device, and obviously it would be bad if any app could just silently unblock itself by sending Rethink a 'stop' intent. [Termux's Tasker plugin uses a permission for this now](https://termux.dev/en/posts/general/2022/02/15/termux-apps-vulnerability-disclosures.html#termuxtasker-privilege-escalation-vulnerability), as does WireGuard.

(Unfortunately, if an actual Android-style permission is used, it looks like Tasker itself - or any other automation app - would have to have its manifest modified to list that permission before it could use the plugin. Tasker's developer seems to have been willing to do this for both Termux and WireGuard, though.)

[Shortcuts would need similar security considerations](https://termux.dev/en/posts/general/2022/02/15/termux-apps-vulnerability-disclosures.html#termuxwidget-privilege-escalation-vulnerability), though in that case a permission probably doesn't make sense because _every single launcher_ would need to be modified to request it…

**Comment by ignoramous** [comment]
A user writes,

> I'd love to be able to control it via Tasker (automation app for android - I'm sure you know it). 
>
> At the least, being able to turn it on/off via tasker would be awesome.  Bring able to do more things (add to a while list?  Switch DNS settings?   etc) would be killer! 
>
> I know some apps (even some VPN apps) handle this through intents.

**Comment by iulko** [comment]
I counted and around 20 issues would be resolved with tasker integration, but I realise that integration would also need new options added according to new people needs.

I dream about changing wireguard endpoint according to wifi or lte connection, but this would require tasker integration on very low level.

**Comment by ignoramous** [comment]
Yep. I am willing to review pull requests.

**Comment by ignoramous** [comment]
@ShantanuVichare has implemented this (I believe not a plugin per se but "Intent"-driven). From their notes in #1925: 

To configure an automation, in Tasker: (Ref: [Reddit thread](https://www.reddit.com/r/Tailscale/comments/141rkyy/tutorial_turn_taiscale_onoff_automatically_using/))
- add a new task, let's call it "RethinkDNS connect"
- in the task, add a "Send intent" action, you can use the search bar to bring it up
- fill in the following fields and leave the rest as default:
  - Package: `com.celzero.bravedns`
  - Class: `com.celzero.bravedns.VpnControlReceiver`
  - Action: `com.celzero.bravedns.intent.action.VPN_START` -> For start up
  - Action `com.celzero.bravedns.intent.action.VPN_STOP` -> For shutdown

To configure an automation, in Automate: (Ref: [Reddit comment](https://www.reddit.com/r/AutomateUser/comments/18311fn/comment/kaonpuv/))
- using the [Broadcast send](https://llamalab.com/automate/doc/block/broadcast_send.html) block

  - Package: `com.celzero.bravedns`
  - Receiver class: `com.celzero.bravedns.VpnControlReceiver`
  - Action: `com.celzero.bravedns.intent.action.VPN_START` -> For start up
  - Action `com.celzero.bravedns.intent.action.VPN_STOP` -> For shutdown

---

Also see additional notes / screenshots in: #2033

---

Thank you @ShantanuVichare 🙏🙂

We owe you one.

---

All: *I'm closing this issue. If there's anything more that is missing with respect to automation / Tasker, let me know what that is, and I'll reopen to track fixing/implementing that.*

**Comment by CheariX** [comment]
This is a really handy addition. Thank you.

What I'm missing is Tasker Wireguard control, similarly to [this](https://openvpn.net/connect-docs/how-to-use-tasker.html).

Currently, we can only start/stop rethink.

**Comment by ignoramous** [comment]
> What I'm missing is Tasker Wireguard control, similarly to [this](https://openvpn.net/connect-docs/how-to-use-tasker.html).

We'll build in SSID-based automation for WireGuard (and DNS, too) into Rethink ourselves: #68 (coming in v055u, the next pre-release version due soon).

If doing it with Tasker / other such automation apps has benefits, let us know and we can track its impl in a new issue.

**Comment by Docga1835** [comment]
I want to add to this. I use MacroDroid (similar to Tasker) to automate my VPNs. I have been able to get it to start and shutdown Rethink correctly. But I need to start and stop the Wireguard based on criteria other than SSID. For instance, I need my VPN to start when I open Thunderbird to write an email and if I open a private tab in my browser. Additionally, having this functionality would allow me to make a quick settings tile that will toggle just the VPN, not start/stop the entire app.

**Comment by ignoramous** [comment]
> Additionally, having this functionality would allow me to make a quick settings tile that will toggle just the VPN, not start/stop the entire app.

Gotcha.

> For instance, I need my VPN to start when I open Thunderbird to write an email

Rethink supports per-app WireGuard.

You can add just *Thunderbird* to a particular WireGuard config (in *Configure -> Proxy -> Setup WireGuard -> *Advanced*), as Rethink supports running multiple WireGuard in parallel. Make sure to mark the WireGuard as *Lockdown* to avoid leaking connections of apps part of it.

> and if I open a private tab in my browser. 

That is not doable but if you use a separate browser altogether, then you could the above (what's mentioned for Thunderbird).

I'll open a new issue to track WireGuard automation.

**Comment by Docga1835** [comment]
> > Additionally, having this functionality would allow me to make a quick settings tile that will toggle just the VPN, not start/stop the entire app.
> 
> Gotcha.

If not adding more Tasker functionality, would you be willing to consider making a quick settings tile for this? This is how the OpenVPN and Wireguard apps' quick settings tiles operate and this is much more preferable to the current quick settings tile. I have a feeling that most users want to use your app as an always on vpn like me and have no need for the current quick settings tile. Also, having the tile automatically open to wireguard list when long-pressed would be extremely helpful as a shortcut to select a particular vpn if needed.

> > For instance, I need my VPN to start when I open Thunderbird to write an email
> 
> Rethink supports per-app WireGuard.
> 
> You can add just _Thunderbird_ to a particular WireGuard config (in *Configure -> Proxy -> Setup WireGuard -> _Advanced_), as Rethink supports running multiple WireGuard in parallel. Make sure to mark the WireGuard as _Lockdown_ to avoid leaking connections of apps part of it.

I will consider doing this but I am only considering the outgoing messages because the ip address is included in the headers. Maybe I can limit it by domain of the outgoing imap server.

> > and if I open a private tab in my browser.
> 
> That is not doable but if you use a separate browser altogether, then you could the above (what's mentioned for Thunderbird).
> 
> I'll open a new issue to track WireGuard automation.

This is possible by intercepting the notification that is created when the notification is generated indicating the private tabs are open. In Macrodroid "notification received" trigger.

**Comment by error-reporting** [comment]
@ignoramous the weird thing is that sending intents without com.celzero.bravedns.VPNControlReceiver receiver class in macrodroid works but as soon as I added the receiver class it doesn't do anything

**Comment by error-reporting** [comment]
> [@ShantanuVichare](https://github.com/ShantanuVichare) has implemented this (I believe not a plugin per se but "Intent"-driven). From their notes in [#1925](https://github.com/celzero/rethink-app/pull/1925):
> 
> To configure an automation, in Tasker: (Ref: [Reddit thread](https://www.reddit.com/r/Tailscale/comments/141rkyy/tutorial_turn_taiscale_onoff_automatically_using/))
> 
> * add a new task, let's call it "RethinkDNS connect"
> * in the task, add a "Send intent" action, you can use the search bar to bring it up
> * fill in the following fields and leave the rest as default:
>   
>   * Package: `com.celzero.bravedns`
>   * Class: `com.celzero.bravedns.VPNControlReceiver`
>   * Action: `com.celzero.bravedns.intent.action.VPN_START` -> For start up
>   * Action `com.celzero.bravedns.intent.action.VPN_STOP` -> For shutdown
> 
> To configure an automation, in Automate: (Ref: [Reddit comment](https://www.reddit.com/r/AutomateUser/comments/18311fn/comment/kaonpuv/))
> 
> * using the [Broadcast send](https://llamalab.com/automate/doc/block/broadcast_send.html) block
>   
>   * Package: `com.celzero.bravedns`
>   * Receiver class: `com.celzero.bravedns.VPNControlReceiver`
>   * Action: `com.celzero.bravedns.intent.action.VPN_START` -> For start up
>   * Action `com.celzero.bravedns.intent.action.VPN_STOP` -> For shutdown
> 
> Also see additional notes / screenshots in: [#2033](https://github.com/celzero/rethink-app/issues/2033)
> 
> Thank you [@ShantanuVichare](https://github.com/ShantanuVichare) 🙏🙂
> 
> We owe you one.
> 
> All: _I'm closing this issue. If there's anything more that is missing with respect to automation / Tasker, let me know what that is, and I'll reopen to track fixing/implementing that._

@ignoramous com.celzero.bravedns.VPNControlReceiver doesn't respond at all and it's com.celzero.bravedns.receiver.VpnControlReceiver that's responsive instead

**Comment by JHBalaji** [comment]
I needed a way to trigger these without Tasker or Termux, so I developed a standalone controller. I use this implementation with Samsung Routines, triggered when I launch certain apps. I have released the source code here.

https://github.com/JHBalaji/rethinkdns-dynamic-shortcuts/releases/tag/v1

**Comment by ignoramous** [comment]
> @ignoramous com.celzero.bravedns.VPNControlReceiver doesn't respond at all and it's com.celzero.bravedns.receiver.VpnControlReceiver that's responsive instead

Changed, thanks.

> I needed a way to trigger these without Tasker or Termux, so I developed a standalone controller.

Nice (: What's the reason for the automation? Rethink, in `v055v`, has auto-start / auto-stop WireGuard based on WiFi SSID / Mobile data.

**Comment by JHBalaji** [comment]
> Nice (: What's the reason for the automation? Rethink, in `v055v`, has auto-start / auto-stop WireGuard based on WiFi SSID / Mobile data.

I need to start and stop the app during launches because some applications do not function well with my firewall and private DNS settings. If the START and STOP intents were exposed as dynamic shortcuts, they could integrate natively with Samsung OneUI routines using IF-THEN-ELSE triggers. I also needed a snooze feature equivalent - disable for X minutes.

**Comment by ignoramous** [comment]
>  I also needed a snooze feature equivalent - disable for X minutes.

Rethink does have "Pause" (on one corner of the START/STOP button), but I don't think it is integrated with automation intents that can be driven by apps like Tasker.

> If the START and STOP intents were exposed as dynamic shortcuts

Inclined to expose these as "dynamic shortcuts" if they're part of Android's SDK (and something Samsung specific). I am not aware of what these are...
